src/libpmemfile-posix/file.c

```diff
@@ -1192,67 +1326,23 @@ _pmemfile_renameat2(PMEMfilepool *pfp,
 
 	struct pmemfile_vinode *src_parent = src.vinode;
 	struct pmemfile_vinode *dst_parent = dst.vinode;
-
-	vinode_rdlock2(src_parent, dst_parent);
-
-	struct pmemfile_dirent *src_dirent =
-			vinode_lookup_dirent_by_name_locked(pfp, src_parent,
-					src.remaining, src_namelen);
-	if (!src_dirent) {
-		error = errno;
-
-		vinode_unlock2(dst_parent, src_parent);
-		goto end_unlocked;
-	}
-
-	src_vinode = inode_ref(pfp, src_dirent->inode, src_parent, NULL,
-			src.remaining, src_namelen);
-	if (!src_vinode) {
-		error = errno;
-		goto end_unlocked;
-	}
-
-	struct pmemfile_dirent *dst_dirent =
-			vinode_lookup_dirent_by_name_locked(pfp, dst_parent,
-					dst.remaining, dst_namelen);
-
-	if (dst_dirent) {
-		dst_vinode = inode_ref(pfp, dst_dirent->inode, dst_parent, NULL,
-				dst.remaining, dst_namelen);
-		if (!dst_vinode) {
-			error = errno;
-			goto end_unlocked;
-		}
-	}
-
-	/* we are going to drop locks, so those pointers will not be valid */
-	src_dirent = NULL;
-	dst_dirent = NULL;
-
-	vinode_unlock2(src_parent, dst_parent);
-
-	/* sort all locks */
 	struct pmemfile_vinode *vinodes[4];
 	size_t vinodes_num;
-	vinode_wrlock4(vinodes, &vinodes_num,
-			src_parent, src_vinode,
-			dst_parent, dst_vinode);
 
-	/* another thread may have modified [src|dst]_parent, refresh */
-	src_dirent = vinode_lookup_dirent_by_name_locked(pfp, src_parent,
-					src.remaining, src_namelen);
-	dst_dirent = vinode_lookup_dirent_by_name_locked(pfp, dst_parent,
-					dst.remaining, dst_namelen);
-
-	/* are [src|dst]_vinode still valid? */
-	if (!src_dirent ||
-			!TOID_EQUALS(src_dirent->inode, src_vinode->tinode) ||
-			(dst_vinode == NULL && dst_dirent != NULL) ||
-			(dst_vinode != NULL && dst_dirent == NULL) ||
-			(dst_vinode != NULL &&
-			!TOID_EQUALS(dst_dirent->inode, dst_vinode->tinode))) {
-		restart = true;
-		goto end_unlock;
+	struct pmemfile_dirent_info src_info, dst_info;
+
+	/*
+	 * lock_parents_and_children can race with another thread messing with
+	 * source or destination directory. Loop as long as race occurs.
+	 */
+	do {
+		error = lock_parents_and_children(pfp, &src, &src_info, &dst,
+				&dst_info, vinodes, &vinodes_num);
+	} while (error == 1);
+
+	if (error < 0) {
+		error = -error;
```

Comment:
Well, if the code for restarting would be negative, and the regular errno numbers would be returned as positive, this would be a tiny bit more simple here.